### ISSUE TITLE
[cilium - eni IPAM MODE]  problem with connection from POD to external resource in additional subnet after kops upgrade from 1.23 -> 1.25 

### ISSUE THREAD

**Issue by glomikuz** [kubernetes/kops#14575]
/kind bug

**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

Version 1.23.2 (git-4125bbbe975ca104c609ceaa625a9a5ff3ac19f4) → Client version: 1.25.2 (git-v1.25.2)

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

1.23.8

Subnets:
  - cidr: 10.210.140.0/26
  - cidr: 10.210.140.64/26
  - cidr: 10.210.0.0/22

**3. What cloud provider are you using?**

AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

kops1.25 replace -f value.yaml
kops1.25 update cluster --name ${CLUSTER_NAME} --yes

**5. What happened after the commands executed?**

https://gist.github.com/glomikuz/5786db5989fa714cf80dd3642b068ee8
Cluster changes have been applied to the cloud.

**6. What did you expect to happen?**

Pods with IPs from my new subnet(  - cidr: 10.210.0.0/22 ) will able connect to external resources

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  name: k8s.tmp-test.test
spec:
  api:
    loadBalancer:
      type: Internal
      useForInternalApi: true
  authorization:
    rbac: {}
  cloudConfig:
    awsEBSCSIDriver:
      enabled: true
  cloudProvider: aws
  clusterAutoscaler:
    enabled: true
  configBase: s3://states/k8s.tmp-test.test
  containerRuntime: containerd
  docker:
    skipInstall: true
  etcdClusters:
  - cpuRequest: 200m
    etcdMembers:
    - instanceGroup: master-1a
      name: a
    manager:
      env:
      - name: ETCD_LISTEN_METRICS_URLS
        value: http://0.0.0.0:2379
      - name: ETCD_METRICS
        value: basic
      - name: ETCD_MANAGER_HOURLY_BACKUPS_RETENTION
        value: 1d
      - name: ETCD_MANAGER_DAILY_BACKUPS_RETENTION
        value: 1d
      - name: ETCD_MAX_REQUEST_BYTES
        value: "6291456"
    memoryRequest: 100Mi
    name: main
  - cpuRequest: 100m
    etcdMembers:
    - instanceGroup: master-1a
      name: a
    manager:
      env:
      - name: ETCD_MANAGER_HOURLY_BACKUPS_RETENTION
        value: 1d
      - name: ETCD_MANAGER_DAILY_BACKUPS_RETENTION
        value: 1d
      - name: ETCD_MAX_REQUEST_BYTES
        value: "6291456"
    memoryRequest: 100Mi
    name: events
  iam:
    legacy: false
  kubeAPIServer:
    featureGates:
      TTLAfterFinished: "true"
    oidcClientID: kubernetes
    oidcGroupsClaim: groups
    oidcIssuerURL: https://sso.test.io/auth/realms/test
  kubeControllerManager:
    featureGates:
      TTLAfterFinished: "true"
  kubeDNS:
    nodeLocalDNS:
      cpuRequest: 25m
      enabled: true
      memoryRequest: 5Mi
    provider: CoreDNS
  kubeProxy:
    enabled: false
  kubelet:
    anonymousAuth: false
    authenticationTokenWebhook: true
    authorizationMode: Webhook
    evictionHard: memory.available<7%,nodefs.available<10%,nodefs.inodesFree<5%,imagefs.available<10%,imagefs.inodesFree<5%
    evictionMaxPodGracePeriod: 30
    evictionSoft: memory.available<12%
    evictionSoftGracePeriod: memory.available=200s
  kubernetesApiAccess:
  - 10.0.0.0/8
  kubernetesVersion: 1.23.8
  masterInternalName: api-internal.k8s.tmp-test.test
  masterPublicName: api.k8s.tmp-test.test
  metricsServer:
    enabled: true
    insecure: true
  networkCIDR: 10.210.0.0/16
  networkID: vpc-ID
  networking:
    cilium:
      enableNodePort: true
      enablePrometheusMetrics: true
      ipam: eni
  nodeProblemDetector:
    enabled: false
  nodeTerminationHandler:
    enabled: true
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 10.0.0.0/8
  subnets:
  - cidr: 10.210.140.0/26
    name: kops-k8s-1a
    type: Private
    zone: eu-central-1a
  - cidr: 10.210.140.64/26
    name: kops-k8s-utility-1a
    type: Utility
    zone: eu-central-1a
  topology:
    dns:
      type: Private
    masters: private
    nodes: private
```
Instance groups
```
apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: k8s.tmp-test
  name: master-1a
spec:
  image: ami-***
  machineType: t3.small
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-1a
  role: Master
  subnets:
  - kops-k8s-1a
 
apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: k8s.tmp-test.test
  name: nodes
spec:
  additionalUserData:
  - content: |
      #!/bin/sh
      mkdir -p /var/lib/test-jvm-dump
    name: add-folder.sh
    type: text/x-shellscript
  image: ami-***
  machineType: t3.small
  maxSize: 2
  minSize: 2
  mixedInstancesPolicy:
    instances:
    - t3.small
    onDemandAboveBase: 0
    onDemandBase: 0
    spotAllocationStrategy: lowest-price
    spotInstancePools: 3
  nodeLabels:
    kops.k8s.io/instancegroup: nodes
  role: Node
  subnets:
  - kops-k8s-1a
  suspendProcesses:
  - AZRebalance
  sysctlParameters:
  - net.netfilter.nf_conntrack_max = 1048576
  - net.core.netdev_max_backlog = 30000
  - net.core.rmem_max = 134217728
  - net.core.wmem_max = 134217728
  - net.ipv4.tcp_wmem = 4096 87380 67108864
  - net.ipv4.tcp_rmem = 4096 87380 67108864
  - net.ipv4.tcp_mem = 187143 249527 1874286
  - net.ipv4.tcp_max_syn_backlog = 8192
  - net.ipv4.ip_local_port_range = 10240 65535
```

**9. Anything else do we need to know?**

Before upgrade all works fine
a get my additional subnet (10.210.0.0/22)  that Cilium using for allocation 

```
kubectl get po  -o wide
NAME                         READY   STATUS              RESTARTS   AGE    IP               NODE
tmp-shell-5b759b78d4-8hbqs   1/1     Running             0          20s   10.210.3.62   ip-10-210-140-26  
tmp-shell-5b759b78d4-dch6v   1/1     Running             0          20s   10.210.0.132  ip-10-210-140-38
tmp-shell-5b759b78d4-dpt7d   1/1     Running             0          20s   10.210.140.37 ip-10-210-140-38  

kubectl exec -it tmp-shell-5b759b78d4-8hbqs bash
ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=51 time=2.04 ms
```

all OK

```
kubectl exec -it tmp-shell-5b759b78d4-dpt7d bash
ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=49 ttl=51 time=1.01 ms
```

all OK

after upgrade to kops 1.25.2

```
kubectl get po  -o wide
NAME                         READY   STATUS              RESTARTS   AGE     IP             NODE
tmp-shell-5b759b78d4-mlxk4   1/1     Running             0          7m45s   10.210.140.46  ip-10-210-140-12
tmp-shell-5b759b78d4-sgf88   1/1     Running             0          7m45s   10.210.1.212   ip-10-210-140-33
tmp-shell-5b759b78d4-smsgq   1/1     Running             0          7m45s   10.210.3.63    ip-10-210-140-12

kubectl exec -it tmp-shell-5b759b78d4-mlxk4 bash
bash-5.2# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=51 time=2.29 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=51 time=1.30 ms
```

ALL OK 

```
kubectl exec -it tmp-shell-5b759b78d4-sgf88 bash
bash-5.2# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
^C
--- 8.8.8.8 ping statistics ---
22 packets transmitted, 0 received, 100% packet loss, time 21494ms
```

NOT OK - we have problem with connect to external host

after update kops created new manifest for cilium

```
ManagedFile/k8s.tmp-test.tmp-addons-networking.cilium.io-k8s-1.16
  	Location            	addons/networking.cilium.io/k8s-1.16-v1.11.yaml

- id: k8s-1.16
                    	+     manifest: networking.cilium.io/k8s-1.16-v1.11.yaml
                    	-     manifest: networking.cilium.io/k8s-1.16-v1.10.yaml
                    	+     manifestHash: 6919848c0feff5b38740fecdeaabe0abc2a7d71f148bfed6e701144fcc99a072
                    	-     manifestHash: cd46e38529278b2b4da5819c9d678c3ba71f7fae43acca2462d9fc5c9742b0bf
                    	      name: networking.cilium.io
                    	      needsRollingUpdate: all
```
and new version of cilium agent/operator
```
image: http://quay.io/cilium/cilium:v1.10.5   -> image: http://quay.io/cilium/cilium:v1.11.8@sha256:7f030cfb166a8aeabcb8208844d96abe5b00a3feb251e8ca47f21cf7b3a74f88
```
cilium manifest [kops1.23-cilium1.10.5](https://gist.github.com/glomikuz/54a6a00dc621f261b15be6709556074a) 
cilium manifest [kops1.25-cilium1.11.8](https://gist.github.com/glomikuz/c8c0cf25193d1f940548d005f0a88b4c)

If i downgrade cilium version -> it will works  

```
  networking:
    cilium:
      ****
      version: v1.10.5
```
Any ideas what i need add to Network configuration ?

PS: same issue when upgrades from 1.23 to 1.24

`Client version: 1.24.4 (git-v1.24.4)`

https://gist.github.com/glomikuz/012abde282948c1ff0037ec0ed1f0ec2

```
k get po -o wide     
NAME                         READY   STATUS    RESTARTS   AGE    IP              NODE                                           
tmp-shell-5b759b78d4-6cmdw   1/1     Running   0          5m1s   10.210.2.74     ip-10-210-140-43.eu-central-1.compute.internal   
tmp-shell-5b759b78d4-fw8sn   1/1     Running   0          5m1s   10.210.2.41     ip-10-210-140-53.eu-central-1.compute.internal   
tmp-shell-5b759b78d4-nvtxv   1/1     Running   0          5m1s   10.210.140.59   ip-10-210-140-53.eu-central-1.compute.internal  


k exec -it tmp-shell-5b759b78d4-fw8sn bash
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
bash-5.2# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
^C
--- 8.8.8.8 ping statistics ---
7 packets transmitted, 0 received, 100% packet loss, time 6126ms
```
NOT OK

```
k exec -it tmp-shell-5b759b78d4-nvtxv bash
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
bash-5.2# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=110 time=2.47 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=110 time=1.76 ms
64 bytes from 8.8.8.8: icmp_seq=3 ttl=110 time=2.05 ms
```
ALL OK with old subnet

**Comment by olemarkus** [comment]
if it works to downgrade cilium, I wonder if this is an upstream bug. I don't think we do any big changes to the manifests that could affect this.

That being said, normally the nonMasqCIDR should match VPC CIDR in ENI mode. Also ensure there is no tunnelling enabled.

**Comment by glomikuz** [comment]
I tried with  nonMasqueradeCIDR: **10.210.0.0/10**  got same behaviour  ( problem with access to external )

but i found some interesting facts 

when we using ENI we must specify masquerade to false 
https://github.com/kubernetes/kops/blob/851fce2b2e24a6b9cdc78122d3743d8ebbfc30d8/k8s/crds/kops.k8s.io_clusters.yaml#L4905-L4912

--masquerade=false by default in cilium manifest - > https://gist.github.com/glomikuz/54a6a00dc621f261b15be6709556074a#file-kops1-23-cilium1-10-5-L62

**BUT** 

looks like this parameters deprecated in 1.11  -> 
https://github.com/cilium/cilium/blob/v1.10.5/daemon/cmd/daemon_main.go#L675-L677
instead we must use  - > 
https://github.com/cilium/cilium/blob/v1.11.8/daemon/cmd/daemon_main.go#L693-L694
https://github.com/cilium/cilium/blob/2bfe4024c2b366fc55200c2c65269545a4410cb5/pkg/option/config.go#L346-L347

if compare old and new logs from cilium agent
https://www.diffchecker.com/9IBJhgmu
we see that masquerade option not using by agent 


if i add cilium parameter --enable-ipv4-masquerade='false'"  direct to config map 
` edit cm cilium-config -n kube-system`

and after reboot all cilium agents 

all works fine 

```
k get po -o wide 
NAME                         READY   STATUS    RESTARTS   AGE     IP              NODE                                             
tmp-shell-5b759b78d4-5glq4   1/1     Running   0          8m37s   10.210.140.30   ip-10-210-140-48.eu-central-1.compute.internal
tmp-shell-5b759b78d4-b2llq   1/1     Running   0          8m37s   10.210.140.36   ip-10-210-140-48.eu-central-1.compute.internal 
tmp-shell-5b759b78d4-bdqrl   1/1     Running   0          8m37s   10.210.0.59     ip-10-210-140-48.eu-central-1.compute.internal
tmp-shell-5b759b78d4-vgd5s   1/1     Running   0          8m37s   10.210.0.71     ip-10-210-140-48.eu-central-1.compute.internal       
```   

```
k exec -it tmp-shell-5b759b78d4-bdqrl bash 
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
bash-5.2# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=51 time=2.41 ms
```
ALL OK 
```
k exec -it tmp-shell-5b759b78d4-5glq4 bash 
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
bash-5.2# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=51 time=1.93 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=51 time=1.37 ms
^C
--- 8.8.8.8 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 1.368/1.649/1.931/0.281 ms
```
ALL OK

**Comment by olemarkus** [comment]
Interesting. Thanks for catching this.

**Comment by olemarkus** [comment]
This should be an easy fix, but I want to spend some time setting up an e2e test for this so that we can catch such changes before it hits users.

**Comment by johngmyers** [comment]
@olemarkus Perhaps we need a `cilium-eni` networking option.

**Comment by olemarkus** [comment]
Yeah that may make sense. It seems to be a popular option, so it would make adoption easier too.

**Comment by olemarkus** [comment]
The only configuration I could make work for cilium 1.11 _and 1.10 is 
```
  enable-ipv4-masquerade: "true"
  egress-masquerade-interfaces: ens5
```

If I disable masqing or if I don't specify the masq interface, it seems incoming packets on ens5 is dropped. I'd especially want to avoid setting the interface name like this as it's distro dependent.

From the flow logs, it seems like traffic is masqed even if masqing is disabled. At least all traffic I could find in the flow logs came from the host IP, not the Pod IP.

**Comment by glomikuz** [comment]
@olemarkus  thanks for fix  

as i understand this fix will be in kops 1.26  ?

**Comment by glomikuz** [comment]
@olemarkus  ping )  how can we test ? or need wait for release 1.26  ?  thanks
